### DISCUSSION TITLE
Pyright auto import

### DISCUSSION THREAD

**Discussion by gsingh93** [joaotavora/eglot#1061] [Q&A] [answered]
I see that [`lsp-pyright`](https://emacs-lsp.github.io/lsp-pyright/) has support for automatically adding imports with `lsp-pyright-auto-import-completions`. Is there any way to do something similar with eglot?

**Comment by rrudakov** [comment, marked as answer]
For me it works without any additional configuration. Do you have any particular problem?


https://user-images.githubusercontent.com/6093590/192224088-cf5f5e73-21ac-433e-bfd3-90b0dfa7558c.mov

**Reply by joaotavora** [reply]
Thanks @rrudakov, for monitoring questions like this.  We need input from people like you as I can't address all issues from such a large ecosystem of language servers.

**Reply by rrudakov** [reply]
No problem! I'm happy to help :)

**Reply by gsingh93** [reply]
Thanks @rrudakov. If you take a look at the screenshots below, you can see that if I type `from types import Dy`, I can autocomplete. But if I just type Dy, I don't get the autocomplete suggestion like shown in your video. Do you have any suggestions for how I can debug this? I have my `company-backends` set to `(company-capf)`.

![image](https://user-images.githubusercontent.com/1012677/192320864-f7bb867b-aeb6-40d0-9ed1-130e71e11977.png)

<img width="351" alt="image" src="https://user-images.githubusercontent.com/1012677/192321364-b314dd49-87a0-47f2-8730-d30a5075890c.png">

**Reply by rrudakov** [reply]
Is eglot running? Could you please try to check what's the value of `completion-at-point-functions` variable? You can use `describe-variable` function or `C-h v` if you use vanilla emacs bindings.

Do you see `eglot-completion-at-point` in the list?

**Reply by gsingh93** [reply]
Yup, I see `(eglot-completion-at-point python-completion-at-point t)`

**Reply by rrudakov** [reply]
Could you try to type `Dyn` then switch to events log buffer (`M-x eglot-events-buffer`) and find request and response for method `:textDocument/completion`?

My request/response pair is the following:
```text
[client-request] (id:6) Mon Sep 26 22:57:57 2022:
(:jsonrpc "2.0" :id 6 :method "textDocument/completion" :params
          (:textDocument
           (:uri "file:///home/rrudakov/tmp/hello.py")
           :position
           (:line 0 :character 3)
           :context
           (:triggerKind 1)))
[server-reply] (id:6) Mon Sep 26 22:57:57 2022:
(:jsonrpc "2.0" :id 6 :result
          (:items
           [(:label "DynamicClassAttribute" :kind 6 :data
                    (:workspacePath "/home/rrudakov/tmp" :filePath "/home/rrudakov/tmp/hello.py" :position
                                    (:line 0 :character 3)
                                    :autoImportText "from types import DynamicClassAttribute" :symbolLabel "DynamicClassAttribute")
                    :sortText "12.9999.DynamicClassAttribute" :detail "Auto-import" :labelDetails
                    (:description "types")
                    :documentation
                    (:kind "plaintext" :value "from types import DynamicClassAttribute")
                    :textEdit
                    (:range
                     (:start
                      (:line 0 :character 0)
                      :end
                      (:line 0 :character 3))
                     :newText "DynamicClassAttribute")
                    :additionalTextEdits
                    [(:range
                      (:start
                       (:line 0 :character 0)
                       :end
                       (:line 0 :character 0))
                      :newText "from types import DynamicClassAttribute\n\n\n")])]
           :isIncomplete t))
```

Also would be useful if you post your emacs config and pyright config (if it exists).

**Reply by gsingh93** [reply]
Sorry, this was my fault. After checking the eglot log, I saw that jedi-langserver was running instead of pyright-langserver, it was a PATH issue.

That being said, what libraries is auto import supposed to work for? For example, if I want to type `os.system` or just `system`, should it be able to add `import os`? That completion (and many others from different libraries) isn't showing up for me.

**Reply by rrudakov** [reply]
I can't tell you for sure, this question should probably be asked to pyright developers.

![image](https://user-images.githubusercontent.com/6093590/192380887-ca43d68a-d1b2-4650-bd5d-801cd3e41f0f.png)

Sometimes I can see a lot of auto-import suggestions from libraries installed in virtualenv, but I don't know how does the pyright decide what to import automatically.
